### PR TITLE
Initial development setup

### LINKED ISSUE #6: Set up code base for development
The current state of the code base is quite basic, and while this is not in and of itself a problem, it should at least try to follow some good practices for Python software development.

Following from [The Hitchhiker's Guide to Python](https://docs.python-guide.org/writing/structure/), the following changes or additions are proposed:
- [x] Rename main.py to palletize.py
- [x] Add a requirements.txt or Pipfile to project root
- [x] Add a test_pylettize.py with rudimentary unit tests
- [x] Add a LICENSE with an appropriate license
- [x] Add a setup.py to enable package installation
- [x] Set up code formatting and static code analysis
- [x] Set up pre-commit hooks

Consequently, the following decisions should be made before taking on this issue:
- Should a requirements.txt or Pipfile be set up
- What testing framework should be used?
- What software license is the most appropriate

Some work left for future issues include:
- A more thorough test suite
- Documentation under some sort of docs directory (as per the aforementioned Hitchhiker's Guide)
- Setting up GitHub workflows for automated testing and other CI niceties
- Maybe utilizing tox to test multiple Python environments

### PR DESCRIPTION
**Pull request by frans-johansson**
Initial work to set up the code base for development. This includes setting up unit testing frameworks, code quality checkers, as well as a proper project structure to at least allow development installation of pylettize via pip.

Closes #6

### CODE REVIEW

**Comment by frans-johansson** [comment]
Considering introducing a Pipfile in addition to requirements.txt, the question is if it would be considered kosher to include both.

**Comment by frans-johansson** [comment]
The way accessing the external palette files is done currently is a bit ugly, but will probably do for now

**Comment by frans-johansson** [comment]
Took the liberty to reformat the code with black and isort right away

**Comment by frans-johansson** [comment]
> The way accessing the external palette files is done currently is a bit ugly, but will probably do for now

The palette files are now accessed via the palettes sub-package via `importlib.resources`. This works since the setup.cfg has been set up to include package data, and the MANIFEST.in file has been set up to include the palette files. Unfortunately, the palettes directory had to be turned into an empty python module for this trick to work, otherwise it was not recognized by `importlib`.

**Comment by frans-johansson** [comment]
> Considering introducing a Pipfile in addition to requirements.txt, the question is if it would be considered kosher to include both.

Only the Pipfile is included now

**Comment by frans-johansson** [comment]
Ready for merge. The final work on this branch is summarized as:
- Setting up a pylettize package which is installable locally with `pip install -e .`
- Setting up a separate tests directory with simple pytest unit tests
- Setting up flake8 with various plugins for code style checks
- Setting up black for formatting
- Using pipenv with a Pipfile for requirement handling during development. This may or may not change depending on future requirements

File: MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1 @@
+include pylettize/palettes/*
```
Comment (frans-johansson) [inline MANIFEST.in:1]: I've seen people add the README and LICENSE here. Should we do that too?

File: pylettize/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+"""Pylettize package."""
+
+from .pylettize import main
+
+__all__ = ["main"]
```
Comment (frans-johansson) [inline pylettize/__init__.py:5]: This is temporary. Will probably re-export more specific things when working on #1

File: pylettize/palettes/__init__.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+The following default palettes are available:
+- gruvbox
+- obama
+- primaries
+
+These are nothing more than text files with RGB hex values, making
+it quite easy to include more palettes to the default library in the future.
+"""
+
+OPTIONS = ["gruvbox", "obama", "primaries"]
```
Comment (frans-johansson) [inline pylettize/palettes/__init__.py:13]: Maybe this doesn't belong here now. My idea was to maybe move logic for reading these default palettes into this module at some point, and that exposing the available options could be useful until then.

File: pylettize/pylettize.py [hunk 4]
```diff
@@ -1,65 +1,76 @@
...
 Maps each pixel in a given image to its nearest neighboring color,
 from a given palette in RGB space in terms of the L2 norm.
 """
+from importlib import resources
+
 import numpy as np
 import numpy.typing as npt
-import skimage.io as io
 import skimage.data as test_images
+import skimage.io as imio
```
Comment (frans-johansson) [inline pylettize/pylettize.py:12]: Renamed this import due to conflicts with the standard library package `io`

File: tests/test_pylettize.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        "#ff0000": np.array([1, 0, 0], dtype=np.float32),
+        "#00ff00": np.array([0, 1, 0], dtype=np.float32),
+        "#0000ff": np.array([0, 0, 1], dtype=np.float32),
+        "#ffff00": np.array([1, 1, 0], dtype=np.float32),
+        "#00ffff": np.array([0, 1, 1], dtype=np.float32),
+        "#ff00ff": np.array([1, 0, 1], dtype=np.float32),
+    }
+
+    for hex, expected_value in to_test.items():
+        np.testing.assert_almost_equal(hex_to_rgb(hex), expected_value)
```
Comment (frans-johansson) [inline tests/test_pylettize.py:22]: This could certainly be parametrized at some point. Maybe this could be handled in some future test-oriented PR.

File: tests/test_pylettize.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    """Ensure computed distance maps are reasonable."""
+    color_a = np.array([1, 0, 0], dtype=np.float32)
+    color_b = np.array([0, 1, 1], dtype=np.float32)
+    image = np.full(shape=(5, 5, 3), fill_value=color_a, dtype=np.float32)
+
+    dist_map_to_a = dist_to_color(image, color_a)
+    dist_map_to_b = dist_to_color(image, color_b)
+
+    np.testing.assert_allclose(dist_map_to_a, 0)
+    np.testing.assert_array_compare(operator.__gt__, dist_map_to_b, 0)
```
Comment (frans-johansson) [inline tests/test_pylettize.py:35]: Maybe this approach to testing would align nicely with the [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) framework. Could be worth investigating at some point in the future.

File: tests/test_pylettize.py (hunk 5 above)
Comment (frans-johansson) [reply tests/test_pylettize.py:22]: See #10

File: tests/test_pylettize.py (hunk 6 above)
Comment (frans-johansson) [reply tests/test_pylettize.py:35]: See #11

File: MANIFEST.in (hunk 1 above)
Comment (tim-jonsson) [reply MANIFEST.in:1]: Should probably be included

File: tests/test_pylettize.py (hunk 5 above)
Comment (tim-jonsson) [reply tests/test_pylettize.py:22]: You could test it more thoroughly by getting predefined color hexes and RGB values of mixed colors. But the edge cases might be sufficient in this case.

File: tests/test_pylettize.py (hunk 6 above)
Comment (tim-jonsson) [reply tests/test_pylettize.py:35]: Could be doable, probably impossible in the case for hex_to_rgb.

**Review by tim-jonsson — APPROVED** [review]
Looks good to me. The comments will be completed on other issues.
